lib/concurrent/executor/executor.rb

```diff
@@ -5,12 +5,75 @@
 module Concurrent
 
   module Executor
+    class AtExitAutoTermination
+      def initialize(enabled = true)
+        @mutex     = Mutex.new
+        @executors = Set.new
+        @enabled   = enabled
+        @installed = false
+      end
+
+      def add(executor)
+        @mutex.synchronize { @executors.add executor }
+      end
+
+      def remove(executor)
+        @mutex.synchronize { @executors.delete executor }
+      end
+
+      def enabled?
+        @mutex.synchronize { @enabled }
+      end
+
+      def enabled=(value)
+        @mutex.synchronize { @enabled = value }
+      end
+
+      def terminate
+        executors = @mutex.synchronize { @executors.to_a if @enabled }
+        if executors
+          executors.each(&:kill) # TODO be gentle first
+          executors.each(&:wait_for_termination)
+          executors.clear
+          true
+        else
+          false
+        end
+      end
+
+      def install
+        @installed ||= begin
+          at_exit { terminate }
+          true
+        end
+        self
+      end
+
+      def executors
+        @mutex.synchronize { @executors.to_a }
+      end
+    end
+
+    AT_EXIT_AUTO_TERMINATION = AtExitAutoTermination.new.install
```

Comment:
Ah sorry about that @jdantonio I forgot about that. Problem is that we cannot use `at_exit` hooks directly because they cannot be removed which results in leeking references. So perhaps we could create `Concurrent.at_exit` interface which will return id which can be used to deregister the hook. It would still has to have a global registry of the hooks but it would be useful API and it would be decoupled from Pools so it's kind a middle ground.